bin/varnishd/cache/cache_backend.c

```diff
@@ -233,8 +234,10 @@ vbe_dir_getfd(VRT_CTX, struct worker *wrk, VCL_BACKEND dir, struct backend *bp,
 			VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
 			VSC_C_main->backend_wait_fail++;
 			cw->cw_state = CW_BE_BUSY;
-		}
-	}
+		} else
+			vbe_connwait_dequeue_locked(bp, cw);
```

Comment:
The problem with dequeuing early here is that racing threads might find the queue empty as a consequence and not exercise the `CW_BE_BUSY` code above, skipping the queue. Which is contrary to the problem you are trying to solve.

Reply:
Also I think `n_conn` is wrong for this case

Reply:
This is not possible, we dequeue and increase the `bp->n_conn` variable while holding the mutex.

If the last entry get dequeued then `BE_BUSY` would return true as n_conn >= max_connections (or not if some fetch completes before the thread grab the mutex, but then there is room for it).

Either way, the only way a thread can skip the queue is iff there is no existing queue AND `bp->n_conn` < max_connections: https://github.com/varnishcache/varnish-cache/pull/4154/files#diff-fe1fb2709016ca15daea3e926b7cec272966ee397c5bafe50d6967023dd9f4a7R218-R219
I do not see how a thread could skip the queue.

Reply:
Before I dive deeper, are you aware that `Lck_CondWaitUntil()` unlocks the mutex while waiting?

Reply:
Sorry @asadsa92 my previous comment was referring to a different scenario. I have deleted it.

So the real problem is that, when the queue goes from one entry to zero, any other racing requests need to queue, otherwise they can overtake the original one and in instead.

Reply:
> Sorry @asadsa92 my previous comment was referring to a different scenario. I have deleted it.

No worries. I would rather get this correct the first time.

> So the real problem is that, when the queue goes from one entry to zero, any other racing requests need to queue, otherwise they can overtake the original one and, for example, reach `VCP_Get()` earlier.

Yes, and we hopefully do that here.

Probes are special as they grab fresh connection, skip the queue and are not counted, but that is kind of expected.
The access to probe is serial, so we never have more than one connection active for each backend.
We could end with many connection in the closing state when the interval is low, but I do not count this as a Varnish problem. The VCL user asked for it.

All other in-core users of connection pool has to go through here.

Reply:
OK, I now think I see why I was wrong. To recapitulate in my own words:
The code before this change needed to keep the queue entry as a "temporary reservation" until the `n_conn++` after the actual connection was made, otherwise the queue skip could actually have happened. But now that `n_conn` is incremented with the dequeue, and only decremented upon error, that scenario can not happen any more, because `BE_BUSY()` will be true.

So I apologize for having taken so long.

What might have contributed to my confusion is that `bp->n_conn++` was moved to `vbe_connwait_dequeue_locked()`. I did see the change, but then apparently forgot about it again when I stared at `vbe_dir_getfd()`, which is evident in my (wrong) comment _Also I think `n_conn` is wrong for this case_.

So, now that there is only one call site left, can we please move the body of `vbe_connwait_dequeue_locked()` in-line again (similar to what you did in https://github.com/varnishcache/varnish-cache/commit/dc013e64315bc3b31e81cf8c00112ecaa35da2f8)? Also, with this change it seems to make no sense to signal when _taking_ a connection, because why would we need to wake up a thread when there are now _less_ connections available?

if you agree, feel free to squash this with your commits:

```diff
diff --git a/bin/varnishd/cache/cache_backend.c b/bin/varnishd/cache/cache_backend.c
index 54ec79ef0..f29994ed0 100644
--- a/bin/varnishd/cache/cache_backend.c
+++ b/bin/varnishd/cache/cache_backend.c
@@ -156,16 +156,6 @@ vbe_connwait_signal_locked(const struct backend *bp)
        }
 }
 
-static void
-vbe_connwait_dequeue_locked(struct backend *bp, struct connwait *cw)
-{
-       Lck_AssertHeld(bp->director->mtx);
-       bp->n_conn++;
-       VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
-       vbe_connwait_signal_locked(bp);
-       cw->cw_state = CW_DEQUEUED;
-}
-
 static void
 vbe_connwait_fini(struct connwait *cw)
 {
@@ -230,13 +220,14 @@ vbe_dir_getfd(VRT_CTX, struct worker *wrk, VCL_BACKEND dir, struct backend *bp,
                            wait_end);
                } while (err == EINTR);
                bp->cw_count--;
+               VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
                if (err != 0 && BE_BUSY(bp)) {
-                       VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
                        VSC_C_main->backend_wait_fail++;
                        cw->cw_state = CW_BE_BUSY;
                } else
-                       vbe_connwait_dequeue_locked(bp, cw);
-       } else if (cw->cw_state == CW_DO_CONNECT)
+                       cw->cw_state = CW_DEQUEUED;
+       }
+       if (cw->cw_state != CW_BE_BUSY)
                bp->n_conn++;
        Lck_Unlock(bp->director->mtx);
 ```

Reply:
> OK, I now think I see why I was wrong. To recapitulate in my own words: The code before this change needed to keep the queue entry as a "temporary reservation" until the `n_conn++` after the actual connection was made, otherwise the queue skip could actually have happened. But now that `n_conn` is incremented with the dequeue, and only decremented upon error, that scenario can not happen any more, because `BE_BUSY()` will be true.
> 
> So I apologize for having taken so long.
> 
> What might have contributed to my confusion is that `bp->n_conn++` was moved to `vbe_connwait_dequeue_locked()`. I did see the change, but then apparently forgot about it again when I stared at `vbe_dir_getfd()`, which is evident in my (wrong) comment _Also I think `n_conn` is wrong for this case_.

I must admit I spent more time to do the implementation due lots of moving parts.
I was working in these code path for different project, and was wondering whether we could simplify. It was then it stroke me that there was a bug here. The test case verified my suspicion.

> So, now that there is only one call site left, can we please move the body of `vbe_connwait_dequeue_locked()` in-line again (similar to what you did in [dc013e6](https://github.com/varnishcache/varnish-cache/commit/dc013e64315bc3b31e81cf8c00112ecaa35da2f8))? Also, with this change it seems to make no sense to signal when _taking_ a connection, because why would we need to wake up a thread when there are now _less_ connections available?

For reference, here is copy of the tree: https://github.com/asadsa92/varnish-cache/commits/backend_conn_queue_fixesv2/

Yes, that was I landed on when doing the fix for the bug: https://github.com/asadsa92/varnish-cache/commit/57f5470a117e5e0fa8aacefc2759334d04477729
This was what led me to simplify the implementation and move it into its own function as we no longer needed the reservation for cw. 

> if you agree, feel free to squash this with your commits:
> 
> ```diff
> diff --git a/bin/varnishd/cache/cache_backend.c b/bin/varnishd/cache/cache_backend.c
> index 54ec79ef0..f29994ed0 100644
> --- a/bin/varnishd/cache/cache_backend.c
> +++ b/bin/varnishd/cache/cache_backend.c
> @@ -156,16 +156,6 @@ vbe_connwait_signal_locked(const struct backend *bp)
>         }
>  }
>  
> -static void
> -vbe_connwait_dequeue_locked(struct backend *bp, struct connwait *cw)
> -{
> -       Lck_AssertHeld(bp->director->mtx);
> -       bp->n_conn++;
> -       VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
> -       vbe_connwait_signal_locked(bp);
> -       cw->cw_state = CW_DEQUEUED;
> -}
> -
>  static void
>  vbe_connwait_fini(struct connwait *cw)
>  {
> @@ -230,13 +220,14 @@ vbe_dir_getfd(VRT_CTX, struct worker *wrk, VCL_BACKEND dir, struct backend *bp,
>                             wait_end);
>                 } while (err == EINTR);
>                 bp->cw_count--;
> +               VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
>                 if (err != 0 && BE_BUSY(bp)) {
> -                       VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
>                         VSC_C_main->backend_wait_fail++;
>                         cw->cw_state = CW_BE_BUSY;
>                 } else
> -                       vbe_connwait_dequeue_locked(bp, cw);
> -       } else if (cw->cw_state == CW_DO_CONNECT)
> +                       cw->cw_state = CW_DEQUEUED;
> +       }
> +       if (cw->cw_state != CW_BE_BUSY)
>                 bp->n_conn++;
>         Lck_Unlock(bp->director->mtx);
> ```

Looks good, I like your proposal
We could even change the state to CW_DEQUEUED earlier (not full diff):

> ```diff
> diff --git a/bin/varnishd/cache/cache_backend.c b/bin/varnishd/cache/cache_backend.c
> index 54ec79ef0..f29994ed0 100644
> --- a/bin/varnishd/cache/cache_backend.c
> +++ b/bin/varnishd/cache/cache_backend.c
> @@ -230,13 +220,14 @@ vbe_dir_getfd(VRT_CTX, struct worker *wrk, VCL_BACKEND dir, struct backend *bp,
>                             wait_end);
>                 } while (err == EINTR);
> +               VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
> +               cw->cw_state = CW_DEQUEUED;
>                 bp->cw_count--;
>                 if (err != 0 && BE_BUSY(bp)) {
> -                       VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
>                         VSC_C_main->backend_wait_fail++;
>                         cw->cw_state = CW_BE_BUSY;
> -                } else
> -                       vbe_connwait_dequeue_locked(bp, cw);
> -       } else if (cw->cw_state == CW_DO_CONNECT)
> +                }
> +       }
> +       if (cw->cw_state != CW_BE_BUSY)
>                 bp->n_conn++;
>         Lck_Unlock(bp->director->mtx);
> ```

Let me know which you prefer, I will then push a commit for it and squash later.

Reply:
Yes, your additional simplification makes sense!

Reply:
@asadsa92 did you maybe forget about this thread?

Reply:
I actually did, sorry about that. Let me redo it.

Reply:
It should be OK now. Not sure how forgot about this, but thanks for the reminder.